File: src/sc_neurocore/neurons/models/hay_l5.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Hay et al. 2011 — Layer 5 thick-tufted pyramidal cell

"""Reduced three-compartment Hay L5 pyramidal-cell model."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal

_STATE_NAMES = ("v_s", "h_na", "n_k", "v_t", "m_ca", "h_ca", "m_ih", "v_a", "ca_a")
_PARAM_NAMES = (
    "g_na",
    "g_k",
    "g_l_s",
    "e_na",
    "e_k",
    "e_l",
    "g_ca_t",
    "g_ih",
    "g_l_t",
    "e_ca",
    "e_ih",
    "g_ca_a",
    "g_kca",
    "g_l_a",
    "g_st",
    "g_ta",
    "p_s",
    "p_t",
    "p_a",
    "ca_decay",
    "f_ca",
    "dt",
    "v_threshold",
    "c_m",
)
_NON_NEGATIVE_PARAMS = (
    "g_na",
    "g_k",
    "g_l_s",
    "g_ca_t",
    "g_ih",
    "g_l_t",
    "g_ca_a",
    "g_kca",
    "g_l_a",
    "g_st",
    "g_ta",
    "f_ca",
)
_STRICTLY_POSITIVE_PARAMS = ("p_s", "p_t", "p_a", "ca_decay", "dt", "c_m")
_N_SUBSTEPS = 4
_State = tuple[float, float, float, float, float, float, float, float, float]


def _safe_exp(value: float) -> float:
    """Return ``exp(value)`` and map overflow to ``math.inf``.

    Parameters
    ----------
    value:
        Exponent to evaluate.

    Returns
    -------
    float
        Exponential value in double precision, or positive infinity when the
        exponent overflows.
    """
    try:
        return math.exp(value)
    except OverflowError:
        return math.inf


@dataclass
class HayL5PyramidalNeuron:
    """Reduced Layer 5 thick-tufted pyramidal-cell model after Hay et al. 2011.

    The maintained production surface is a compact three-compartment reduction
    with soma, apical trunk, and apical tuft voltages plus six gates/calcium
    state variables. It preserves the public dual-input API
    ``step(current_soma, current_tuft=0.0)`` while moving the default numerical
    path to candidate-first RK4. The historical explicit Euler path remains
    available through ``integrator="baseline_euler"`` for regression and
    benchmark comparisons.

    Reference: Hay, E. et al. (2011). PLoS Comput. Biol. 7:e1002107.
    """

    v_s: float = -75.0
    h_na: float = 0.9
    n_k: float = 0.1
    v_t: float = -75.0
    m_ca: float = 0.0
    h_ca: float = 1.0
    m_ih: float = 0.0
    v_a: float = -75.0
    ca_a: float = 0.0001
    g_na: float = 300.0
    g_k: float = 40.0
    g_l_s: float = 0.03
    e_na: float = 50.0
    e_k: float = -85.0
    e_l: float = -75.0
    g_ca_t: float = 2.0
    g_ih: float = 0.02
    g_l_t: float = 0.03
    e_ca: float = 140.0
    e_ih: float = -45.0
    g_ca_a: float = 1.5
    g_kca: float = 2.5
    g_l_a: float = 0.03
    g_st: float = 1.5
    g_ta: float = 0.8
    p_s: float = 0.15
    p_t: float = 0.25
    p_a: float = 0.60
    ca_decay: float = 200.0
    f_ca: float = 0.0002
    dt: float = 0.025
    v_threshold: float = -30.0
    c_m: float = 1.0
    integrator: Literal["rk4", "baseline_euler"] = "rk4"

    def __post_init__(self) -> None:
        """Validate the selected integrator and numeric configuration."""
        if self.integrator not in {"rk4", "baseline_euler"}:
            raise ValueError(f"Unsupported integrator for HayL5PyramidalNeuron: {self.integrator}")
        self._validate_configuration()

    @staticmethod
    def _finite_float(name: str, value: float) -> float:
        """Return ``value`` as a finite float and reject booleans.

        Parameters
        ----------
        name:
            Field name used in error messages.
        value:
            Candidate numeric value.

        Returns
        -------
        float
            Finite floating-point value.
        """
        if isinstance(value, bool):
            raise ValueError(f"{name} must be finite")
        try:
            result = float(value)
        except (TypeError, ValueError) as exc:
            raise ValueError(f"{name} must be finite") from exc
        if not math.isfinite(result):
            raise ValueError(f"{name} must be finite")
        return result

    def _validate_configuration(self) -> None:
        """Coerce state and parameters to finite floats and enforce signs."""
        for name in (*_STATE_NAMES, *_PARAM_NAMES):
            setattr(self, name, self._finite_float(name, getattr(self, name)))
        for name in _STRICTLY_POSITIVE_PARAMS:
            if getattr(self, name) <= 0.0:
                raise ValueError(f"{name} must be positive")
        for name in _NON_NEGATIVE_PARAMS:
            if getattr(self, name) < 0.0:
                raise ValueError(f"{name} must be non-negative")
        if self.ca_a < 0.0:
            raise ValueError("ca_a must be non-negative")

    def _validate_runtime_configuration(self) -> None:
        """Re-check state and parameter validity before mutating state."""
        for name in (*_STATE_NAMES, *_PARAM_NAMES):
            self._finite_float(name, getattr(self, name))
        for name in _STRICTLY_POSITIVE_PARAMS:
            if getattr(self, name) <= 0.0:
                raise ValueError(f"{name} must be positive")
        for name in _NON_NEGATIVE_PARAMS:
            if getattr(self, name) < 0.0:
                raise ValueError(f"{name} must be non-negative")
        if self.ca_a < 0.0:
            raise ValueError("ca_a must be non-negative")

    def _derivatives(self, state: _State, current_soma: float, current_tuft: float) -> _State:
        """Return derivatives for all nine state variables.

        Parameters
        ----------
        state:
            Candidate state in ``(v_s, h_na, n_k, v_t, m_ca, h_ca, m_ih, v_a,
            ca_a)`` order.
        current_soma:
            External somatic current.
        current_tuft:
            External apical-tuft current.

        Returns
        -------
        tuple of float
            Nine derivatives in state order.
        """
        if not all(math.isfinite(value) for value in (*state, current_soma, current_tuft)):
            raise FloatingPointError("Hay L5 derivative input became non-finite")
        v_s, h_na, n_k, v_t, m_ca, h_ca, m_ih, v_a, ca_a = state
        ca_eff = max(ca_a, 0.0)

        m_na_inf = 1.0 / (1.0 + _safe_exp(-(v_s + 38.0) / 7.0))
        h_na_inf = 1.0 / (1.0 + _safe_exp((v_s + 65.0) / 6.0))
        n_k_inf = 1.0 / (1.0 + _safe_exp(-(v_s + 25.0) / 12.0))
        tau_h = 0.5 + 14.0 / (1.0 + _safe_exp((v_s + 35.0) / 10.0))
        tau_n = 1.0 + 5.0 / (1.0 + _safe_exp((v_s + 30.0) / 10.0))
        d_h_na = (h_na_inf - h_na) / tau_h
        d_n_k = (n_k_inf - n_k) / tau_n
        i_na = self.g_na * m_na_inf * m_na_inf * m_na_inf * h_na * (v_s - self.e_na)
        i_k = self.g_k * n_k * n_k * n_k * n_k * (v_s - self.e_k)
        i_l_s = self.g_l_s * (v_s - self.e_l)
        i_st = self.g_st * (v_s - v_t) / self.p_s

        m_ca_inf = 1.0 / (1.0 + _safe_exp(-(v_t + 27.0) / 7.0))
        h_ca_inf = 1.0 / (1.0 + _safe_exp((v_t + 52.0) / 5.0))
        m_ih_inf = 1.0 / (1.0 + _safe_exp((v_t + 75.0) / 5.5))
        d_m_ca = m_ca_inf - m_ca
        d_h_ca = (h_ca_inf - h_ca) / 20.0
        d_m_ih = (m_ih_inf - m_ih) / 50.0
        i_ca_t = self.g_ca_t * m_ca * m_ca * h_ca * (v_t - self.e_ca)
        i_ih = self.g_ih * m_ih * (v_t - self.e_ih)
        i_l_t = self.g_l_t * (v_t - self.e_l)
        i_ts = self.g_st * (v_t - v_s) / self.p_t
        i_ta = self.g_ta * (v_t - v_a) / self.p_t

        m_ca_a_inf = 1.0 / (1.0 + _safe_exp(-(v_a + 30.0) / 5.0))
        kca_act = ca_eff / (ca_eff + 0.001)
        i_ca_a = self.g_ca_a * m_ca_a_inf * m_ca_a_inf * (v_a - self.e_ca)
        i_kca = self.g_kca * kca_act * (v_a - self.e_k)
        i_l_a = self.g_l_a * (v_a - self.e_l)
        i_at = self.g_ta * (v_a - v_t) / self.p_a

        d_v_s = (-i_na - i_k - i_l_s - i_st + current_soma / self.p_s) / self.c_m
        d_v_t = (-i_ca_t - i_ih - i_l_t - i_ts - i_ta) / self.c_m
        d_v_a = (-i_ca_a - i_kca - i_l_a - i_at + current_tuft / self.p_a) / self.c_m
        d_ca_a = -self.f_ca * i_ca_a - ca_eff / self.ca_decay

        derivatives = (d_v_s, d_h_na, d_n_k, d_v_t, d_m_ca, d_h_ca, d_m_ih, d_v_a, d_ca_a)
        if not all(math.isfinite(value) for value in derivatives):
            raise FloatingPointError("Hay L5 derivative became non-finite")
        return derivatives

    def _rk4_substep(self, state: _State, current_soma: float, current_tuft: float) -> _State:
        """Return one RK4 candidate for the nine-state vector.

        Parameters
        ----------
        state:
            State at substep start.
        current_soma:
            Somatic current held constant across the RK4 stages.
        current_tuft:
            Tuft current held constant across the RK4 stages.

        Returns
        -------
        tuple of float
            Candidate state after one internal substep.
        """
        dt = self.dt
        k1 = self._derivatives(state, current_soma, current_tuft)
        s2 = (
            state[0] + 0.5 * dt * k1[0],
            state[1] + 0.5 * dt * k1[1],
            state[2] + 0.5 * dt * k1[2],
            state[3] + 0.5 * dt * k1[3],
            state[4] + 0.5 * dt * k1[4],
            state[5] + 0.5 * dt * k1[5],
            state[6] + 0.5 * dt * k1[6],
            state[7] + 0.5 * dt * k1[7],
            state[8] + 0.5 * dt * k1[8],
        )
        k2 = self._derivatives(s2, current_soma, current_tuft)
        s3 = (
            state[0] + 0.5 * dt * k2[0],
            state[1] + 0.5 * dt * k2[1],
            state[2] + 0.5 * dt * k2[2],
            state[3] + 0.5 * dt * k2[3],
            state[4] + 0.5 * dt * k2[4],
            state[5] + 0.5 * dt * k2[5],
            state[6] + 0.5 * dt * k2[6],
            state[7] + 0.5 * dt * k2[7],
            state[8] + 0.5 * dt * k2[8],
        )
        k3 = self._derivatives(s3, current_soma, current_tuft)
        s4 = (
            state[0] + dt * k3[0],
            state[1] + dt * k3[1],
            state[2] + dt * k3[2],
            state[3] + dt * k3[3],
            state[4] + dt * k3[4],
            state[5] + dt * k3[5],
            state[6] + dt * k3[6],
            state[7] + dt * k3[7],
            state[8] + dt * k3[8],
        )
        k4 = self._derivatives(s4, current_soma, current_tuft)
        return (
            state[0] + dt * (k1[0] + 2.0 * k2[0] + 2.0 * k3[0] + k4[0]) / 6.0,
            state[1] + dt * (k1[1] + 2.0 * k2[1] + 2.0 * k3[1] + k4[1]) / 6.0,
            state[2] + dt * (k1[2] + 2.0 * k2[2] + 2.0 * k3[2] + k4[2]) / 6.0,
            state[3] + dt * (k1[3] + 2.0 * k2[3] + 2.0 * k3[3] + k4[3]) / 6.0,
            state[4] + dt * (k1[4] + 2.0 * k2[4] + 2.0 * k3[4] + k4[4]) / 6.0,
            state[5] + dt * (k1[5] + 2.0 * k2[5] + 2.0 * k3[5] + k4[5]) / 6.0,
            state[6] + dt * (k1[6] + 2.0 * k2[6] + 2.0 * k3[6] + k4[6]) / 6.0,
            state[7] + dt * (k1[7] + 2.0 * k2[7] + 2.0 * k3[7] + k4[7]) / 6.0,
            state[8] + dt * (k1[8] + 2.0 * k2[8] + 2.0 * k3[8] + k4[8]) / 6.0,
        )

    def _euler_substep(self, state: _State, current_soma: float, current_tuft: float) -> _State:
        """Return one explicit Euler candidate for regression comparison.

        Parameters
        ----------
        state:
            State at substep start.
        current_soma:
            External somatic current.
        current_tuft:
            External apical-tuft current.

        Returns
        -------
        tuple of float
            Candidate state after one internal substep.
        """
        derivatives = self._derivatives(state, current_soma, current_tuft)
        return (
            state[0] + self.dt * derivatives[0],
            state[1] + self.dt * derivatives[1],
            state[2] + self.dt * derivatives[2],
            state[3] + self.dt * derivatives[3],
            state[4] + self.dt * derivatives[4],
            state[5] + self.dt * derivatives[5],
            state[6] + self.dt * derivatives[6],
            state[7] + self.dt * derivatives[7],
            state[8] + self.dt * derivatives[8],
        )

    @staticmethod
    def _validate_candidate(candidate: _State) -> _State:
        """Return a finite candidate with non-negative calcium or raise.

        Parameters
        ----------
        candidate:
            Candidate state in model state order.

        Returns
        -------
        tuple of float
            Candidate state with tuft calcium clipped to zero when needed.
        """
        if not all(math.isfinite(value) for value in candidate):
            raise FloatingPointError("Hay L5 candidate became non-finite")
        return (
            candidate[0],
            candidate[1],
            candidate[2],
            candidate[3],
            candidate[4],
            candidate[5],
            candidate[6],
            candidate[7],
            max(candidate[8], 0.0),
        )

    def step(self, current_soma: float, current_tuft: float = 0.0) -> int:
        """Advance the model by one public step and return a spike indicator.

        Parameters
        ----------
        current_soma:
            External current delivered to the somatic compartment.
        current_tuft:
            External current delivered to the apical-tuft compartment.

        Returns
        -------
        int
            ``1`` on an upward soma-threshold crossing, otherwise ``0``.
        """
        current_soma = self._finite_float("current_soma", current_soma)
        current_tuft = self._finite_float("current_tuft", current_tuft)
        self._validate_runtime_configuration()
        v_s_prev = self.v_s
        state = (
            self.v_s,
            self.h_na,
            self.n_k,
            self.v_t,
            self.m_ca,
            self.h_ca,
            self.m_ih,
            self.v_a,
            self.ca_a,
        )
        advance = self._euler_substep if self.integrator == "baseline_euler" else self._rk4_substep
        for _ in range(_N_SUBSTEPS):
            state = self._validate_candidate(advance(state, current_soma, current_tuft))

        (
            self.v_s,
            self.h_na,
            self.n_k,
            self.v_t,
            self.m_ca,
            self.h_ca,
            self.m_ih,
            self.v_a,
            self.ca_a,
        ) = state
        return int(self.v_s >= self.v_threshold and v_s_prev < self.v_threshold)

    def reset(self) -> None:
        """Restore voltage, gate, and calcium state to documented defaults."""
        self.v_s = -75.0
        self.h_na = 0.9
        self.n_k = 0.1
        self.v_t = -75.0
        self.m_ca = 0.0
        self.h_ca = 1.0
        self.m_ih = 0.0
        self.v_a = -75.0
        self.ca_a = 0.0001

File: engine/src/neurons/multi_compartment/hay_l5.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Hay layer-5 pyramidal neuron model

//! Hay layer-5 pyramidal neuron model.

/// Hay et al. 2011 — Layer 5 thick-tufted pyramidal (3-compartment reduced).
#[derive(Clone, Debug)]
pub struct HayL5PyramidalNeuron {
    pub v_s: f64,
    pub v_t: f64,
    pub v_a: f64,
    pub h_na: f64,
    pub n_k: f64,
    pub m_ca: f64,
    pub h_ca: f64,
    pub m_ih: f64,
    pub ca_a: f64,
    pub g_na: f64,
    pub g_k: f64,
    pub g_l_s: f64,
    pub g_ca_t: f64,
    pub g_ih: f64,
    pub g_l_t: f64,
    pub g_ca_a: f64,
    pub g_kca: f64,
    pub g_l_a: f64,
    pub g_st: f64,
    pub g_ta: f64,
    pub p_s: f64,
    pub p_t: f64,
    pub p_a: f64,
    pub e_na: f64,
    pub e_k: f64,
    pub e_ca: f64,
    pub e_ih: f64,
    pub e_l: f64,
    pub ca_decay: f64,
    pub f_ca: f64,
    pub c_m: f64,
    pub dt: f64,
    pub v_threshold: f64,
}

impl HayL5PyramidalNeuron {
    pub fn new() -> Self {
        Self {
            v_s: -75.0,
            v_t: -75.0,
            v_a: -75.0,
            h_na: 0.9,
            n_k: 0.1,
            m_ca: 0.0,
            h_ca: 1.0,
            m_ih: 0.0,
            ca_a: 0.0001,
            g_na: 300.0,
            g_k: 40.0,
            g_l_s: 0.03,
            g_ca_t: 2.0,
            g_ih: 0.02,
            g_l_t: 0.03,
            g_ca_a: 1.5,
            g_kca: 2.5,
            g_l_a: 0.03,
            g_st: 1.5,
            g_ta: 0.8,
            p_s: 0.15,
            p_t: 0.25,
            p_a: 0.60,
            e_na: 50.0,
            e_k: -85.0,
            e_ca: 140.0,
            e_ih: -45.0,
            e_l: -75.0,
            ca_decay: 200.0,
            f_ca: 0.0002,
            c_m: 1.0,
            dt: 0.025,
            v_threshold: -30.0,
        }
    }
    fn valid(&self) -> bool {
        self.v_s.is_finite()
            && self.h_na.is_finite()
            && self.n_k.is_finite()
            && self.v_t.is_finite()
            && self.m_ca.is_finite()
            && self.h_ca.is_finite()
            && self.m_ih.is_finite()
            && self.v_a.is_finite()
            && self.ca_a.is_finite()
            && self.ca_a >= 0.0
            && self.g_na.is_finite()
            && self.g_na >= 0.0
            && self.g_k.is_finite()
            && self.g_k >= 0.0
            && self.g_l_s.is_finite()
            && self.g_l_s >= 0.0
            && self.g_ca_t.is_finite()
            && self.g_ca_t >= 0.0
            && self.g_ih.is_finite()
            && self.g_ih >= 0.0
            && self.g_l_t.is_finite()
            && self.g_l_t >= 0.0
            && self.g_ca_a.is_finite()
            && self.g_ca_a >= 0.0
            && self.g_kca.is_finite()
            && self.g_kca >= 0.0
            && self.g_l_a.is_finite()
            && self.g_l_a >= 0.0
            && self.g_st.is_finite()
            && self.g_st >= 0.0
            && self.g_ta.is_finite()
            && self.g_ta >= 0.0
            && self.p_s.is_finite()
            && self.p_s > 0.0
            && self.p_t.is_finite()
            && self.p_t > 0.0
            && self.p_a.is_finite()
            && self.p_a > 0.0
            && self.e_na.is_finite()
            && self.e_k.is_finite()
            && self.e_ca.is_finite()
            && self.e_ih.is_finite()
            && self.e_l.is_finite()
            && self.ca_decay.is_finite()
            && self.ca_decay > 0.0
            && self.f_ca.is_finite()
            && self.f_ca >= 0.0
            && self.c_m.is_finite()
            && self.c_m > 0.0
            && self.dt.is_finite()
            && self.dt > 0.0
            && self.v_threshold.is_finite()
    }

    fn derivatives(&self, s: [f64; 9], current_soma: f64, current_tuft: f64) -> [f64; 9] {
        let v_s = s[0];
        let h_na = s[1];
        let n_k = s[2];
        let v_t = s[3];
        let m_ca = s[4];
        let h_ca = s[5];
        let m_ih = s[6];
        let v_a = s[7];
        let ca_a = s[8].max(0.0);

        let m_na_inf = 1.0 / (1.0 + (-(v_s + 38.0) / 7.0).exp());
        let h_na_inf = 1.0 / (1.0 + ((v_s + 65.0) / 6.0).exp());
        let n_k_inf = 1.0 / (1.0 + (-(v_s + 25.0) / 12.0).exp());
        let tau_h = 0.5 + 14.0 / (1.0 + ((v_s + 35.0) / 10.0).exp());
        let tau_n = 1.0 + 5.0 / (1.0 + ((v_s + 30.0) / 10.0).exp());
        let i_na = self.g_na * m_na_inf * m_na_inf * m_na_inf * h_na * (v_s - self.e_na);
        let i_k = self.g_k * n_k * n_k * n_k * n_k * (v_s - self.e_k);
        let i_l_s = self.g_l_s * (v_s - self.e_l);
        let i_st = self.g_st * (v_s - v_t) / self.p_s;

        let m_ca_inf = 1.0 / (1.0 + (-(v_t + 27.0) / 7.0).exp());
        let h_ca_inf = 1.0 / (1.0 + ((v_t + 52.0) / 5.0).exp());
        let m_ih_inf = 1.0 / (1.0 + ((v_t + 75.0) / 5.5).exp());
        let i_ca_t = self.g_ca_t * m_ca * m_ca * h_ca * (v_t - self.e_ca);
        let i_ih = self.g_ih * m_ih * (v_t - self.e_ih);
        let i_l_t = self.g_l_t * (v_t - self.e_l);
        let i_ts = self.g_st * (v_t - v_s) / self.p_t;
        let i_ta = self.g_ta * (v_t - v_a) / self.p_t;

        let m_ca_a_inf = 1.0 / (1.0 + (-(v_a + 30.0) / 5.0).exp());
        let kca_act = ca_a / (ca_a + 0.001);
        let i_ca_a = self.g_ca_a * m_ca_a_inf * m_ca_a_inf * (v_a - self.e_ca);
        let i_kca = self.g_kca * kca_act * (v_a - self.e_k);
        let i_l_a = self.g_l_a * (v_a - self.e_l);
        let i_at = self.g_ta * (v_a - v_t) / self.p_a;

        [
            (-i_na - i_k - i_l_s - i_st + current_soma / self.p_s) / self.c_m,
            (h_na_inf - h_na) / tau_h,
            (n_k_inf - n_k) / tau_n,
            (-i_ca_t - i_ih - i_l_t - i_ts - i_ta) / self.c_m,
            m_ca_inf - m_ca,
            (h_ca_inf - h_ca) / 20.0,
            (m_ih_inf - m_ih) / 50.0,
            (-i_ca_a - i_kca - i_l_a - i_at + current_tuft / self.p_a) / self.c_m,
            -self.f_ca * i_ca_a - ca_a / self.ca_decay,
        ]
    }

    fn rk4_substep(&self, s: [f64; 9], current_soma: f64, current_tuft: f64) -> [f64; 9] {
        let dt = self.dt;
        let k1 = self.derivatives(s, current_soma, current_tuft);
        let mut s2 = [0.0; 9];
        let mut s3 = [0.0; 9];
        let mut s4 = [0.0; 9];
        for i in 0..9 {
            s2[i] = s[i] + 0.5 * dt * k1[i];
        }
        let k2 = self.derivatives(s2, current_soma, current_tuft);
        for i in 0..9 {
            s3[i] = s[i] + 0.5 * dt * k2[i];
        }
        let k3 = self.derivatives(s3, current_soma, current_tuft);
        for i in 0..9 {
            s4[i] = s[i] + dt * k3[i];
        }
        let k4 = self.derivatives(s4, current_soma, current_tuft);
        let mut next = [0.0; 9];
        for i in 0..9 {
            next[i] = s[i] + dt * (k1[i] + 2.0 * k2[i] + 2.0 * k3[i] + k4[i]) / 6.0;
        }
        next[8] = next[8].max(0.0);
        next
    }

    pub fn step(&mut self, current_soma: f64, current_tuft: f64) -> i32 {
        if !current_soma.is_finite() || !current_tuft.is_finite() || !self.valid() {
            return 0;
        }
        let v_s_prev = self.v_s;
        let mut state = [
            self.v_s, self.h_na, self.n_k, self.v_t, self.m_ca, self.h_ca, self.m_ih, self.v_a,
            self.ca_a,
        ];
        for _ in 0..4 {
            state = self.rk4_substep(state, current_soma, current_tuft);
            if !state.iter().all(|value| value.is_finite()) {
                return 0;
            }
        }
        self.v_s = state[0];
        self.h_na = state[1];
        self.n_k = state[2];
        self.v_t = state[3];
        self.m_ca = state[4];
        self.h_ca = state[5];
        self.m_ih = state[6];
        self.v_a = state[7];
        self.ca_a = state[8];
        if self.v_s >= self.v_threshold && v_s_prev < self.v_threshold {
            1
        } else {
            0
        }
    }
    pub fn reset(&mut self) {
        self.v_s = -75.0;
        self.v_t = -75.0;
        self.v_a = -75.0;
        self.h_na = 0.9;
        self.n_k = 0.1;
        self.m_ca = 0.0;
        self.h_ca = 1.0;
        self.m_ih = 0.0;
        self.ca_a = 0.0001;
    }
}
impl Default for HayL5PyramidalNeuron {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hay_fires() {
        let mut n = HayL5PyramidalNeuron::new();
        let t: i32 = (0..500).map(|_| n.step(20.0, 0.0)).sum();
        assert!(t > 0);
    }

    #[test]
    fn hay_reset() {
        let mut n = HayL5PyramidalNeuron::new();
        for _ in 0..100 {
            n.step(20.0, 0.0);
        }
        n.reset();
        assert!((n.v_s - (-75.0)).abs() < 1e-10);
    }

    #[test]
    fn hay_bounded() {
        let mut n = HayL5PyramidalNeuron::new();
        for _ in 0..500 {
            n.step(100.0, 0.0);
        }
        assert!(n.v_s.is_finite());
    }

    #[test]
    fn hay_nan_no_panic() {
        HayL5PyramidalNeuron::new().step(f64::NAN, 0.0);
    }

    #[test]
    fn hay_rk4_somatic_anchor() {
        let mut n = HayL5PyramidalNeuron::new();
        let spikes: i32 = (0..20_000).map(|_| n.step(10.0, 0.0)).sum();
        assert_eq!(spikes, 1);
        assert!(n.ca_a >= 0.0);
    }

    #[test]
    fn hay_rk4_dual_input_anchor() {
        let mut n = HayL5PyramidalNeuron::new();
        let spikes: i32 = (0..20_000).map(|_| n.step(5.0, 5.0)).sum();
        assert_eq!(spikes, 4);
    }

    #[test]
    fn hay_invalid_input_preserves_state() {
        let mut n = HayL5PyramidalNeuron::new();
        for _ in 0..10 {
            n.step(10.0, 0.0);
        }
        let old = [
            n.v_s, n.h_na, n.n_k, n.v_t, n.m_ca, n.h_ca, n.m_ih, n.v_a, n.ca_a,
        ];
        assert_eq!(n.step(f64::INFINITY, 0.0), 0);
        assert_eq!(
            [n.v_s, n.h_na, n.n_k, n.v_t, n.m_ca, n.h_ca, n.m_ih, n.v_a, n.ca_a],
            old
        );
    }
}
